### PR TITLE
datapath: Fix security ID propagation in tunnel header for NodePort BPF forwarded requests

### LINKED ISSUE #19010: inconsistent result with cilium network policies when kube-proxy is enabled or disabled
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

I am running cilium with kube-proxy disabled(installed using rke2-1.20.12).
```
root@a172-25-172-204:/home/cilium# cilium status
KVStore:                Ok   Disabled
Kubernetes:             Ok   1.20 (v1.20.12+rke2r1) [linux/amd64]
Kubernetes APIs:        ["cilium/v2::CiliumClusterwideNetworkPolicy", "cilium/v2::CiliumEndpoint", "cilium/v2::CiliumNetworkPolicy", "cilium/v2::CiliumNode", "core/v1::Namespace", "core/v1::Node", "core/v1::Pods", "core/v1::Service", "discovery/v1beta1::EndpointSlice", "networking.k8s.io/v1::NetworkPolicy"]
KubeProxyReplacement:   Strict   [eth0 (Direct Routing)]
Cilium:                 Ok       OK
NodeMonitor:            Listening for events on 4 CPUs with 64x4096 of shared memory
Cilium health daemon:   Ok
IPAM:                   IPv4: 23/255 allocated from 172.23.128.0/24,
BandwidthManager:       Disabled
Host Routing:           Legacy
Masquerading:           BPF   [eth0]   172.23.128.0/24
Controller Status:      102/102 healthy
Proxy Status:           OK, ip 172.23.128.26, 0 redirects active on ports 10000-20000
Hubble:                 Ok              Current/Max Flows: 4096/4096 (100.00%), Flows/s: 34.77   Metrics: Disabled
Cluster health:         4/4 reachable   (2022-03-02T05:54:47Z)
root@a172-25-172-204:/home/cilium#

root@172.25.172.204:~# kubectl get pods -A | grep kube-proxy
root@172.25.172.204:~#
```

I have a global clusterwide deny policy which blocks all ingress to cluster but allows ingress from cluster entities.
```
---
apiVersion: "cilium.io/v2"
kind: CiliumClusterwideNetworkPolicy
metadata:
  name: "global-deny-all"
spec:
  description: "block everything coming from outside the cluster"
  endpointSelector: {}
  ingress:
  - fromEntities:
    - cluster
```
I deployed an app, created a service and exposed it using type as loadBalancer.
```
root@172.25.172.204:~# kubectl get svc -n web
NAME    TYPE           CLUSTER-IP      EXTERNAL-IP      PORT(S)          AGE
nginx   LoadBalancer   172.23.60.147   172.25.172.219   4080:32129/TCP   5h52m
root@172.25.172.204:~#
```
I opened access to the app from only one IP address by adding the CNP.
```
apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
metadata:
  name: "allow-nginx"
  namespace: web
spec:
  endpointSelector:
    matchLabels:
      app: nginx
  ingress:
    - fromCIDR:
      - 198.18.135.71/32
```
I am expecting that if I try to access the app from a machine outside the cluster different than the allowed machine, it should be blocked. However, I am seeing it getting allowed. My hubble logs are:
```
root@172.25.172.204:~# kubectl -n kube-system exec cilium-2rp4t -- hubble observe --since 3m --pod web/nginx-695fc7b69-jv4q7
TIMESTAMP             SOURCE                         DESTINATION                    TYPE          VERDICT     SUMMARY
Mar  2 05:49:22.550   172.23.129.66:55166            web/nginx-695fc7b69-jv4q7:80   L3-Only       FORWARDED   TCP Flags: SYN
Mar  2 05:49:22.550   172.23.129.66:55166            web/nginx-695fc7b69-jv4q7:80   to-endpoint   FORWARDED   TCP Flags: SYN
Mar  2 05:49:22.550   web/nginx-695fc7b69-jv4q7:80   172.23.129.66:55166            to-overlay    FORWARDED   TCP Flags: SYN, ACK
Mar  2 05:49:22.553   172.23.129.66:55166            web/nginx-695fc7b69-jv4q7:80   to-endpoint   FORWARDED   TCP Flags: ACK
Mar  2 05:49:22.553   172.23.129.66:55166            web/nginx-695fc7b69-jv4q7:80   to-endpoint   FORWARDED   TCP Flags: ACK, PSH
Mar  2 05:49:22.553   web/nginx-695fc7b69-jv4q7:80   172.23.129.66:55166            to-overlay    FORWARDED   TCP Flags: ACK, PSH
Mar  2 05:49:22.556   172.23.129.66:55166            web/nginx-695fc7b69-jv4q7:80   to-endpoint   FORWARDED   TCP Flags: ACK, FIN
Mar  2 05:49:22.556   web/nginx-695fc7b69-jv4q7:80   172.23.129.66:55166            to-overlay    FORWARDED   TCP Flags: ACK, FIN
Mar  2 05:49:22.558   172.23.129.66:55166            web/nginx-695fc7b69-jv4q7:80   to-endpoint   FORWARDED   TCP Flags: ACK
```
I sometimes see the policy getting applied correctly, but sometimes it doesn't work as expected. Sometimes I see this issue when I modify the CNP or re-schedule pods on other nodes. I start seeing the unexpected behavior and not sure what I might be missing here.

I had this working fine with kube-proxy enabled. Global deny all always worked as expected. With kube-proxy, I had to set `externalTrafficPolicy: local` in service to avoid SNAT and have client IP preserved when evaluating the CNP. I am trying to test it with kube-proxy disabled now to see if CNP works there without requiring the externalTrafficPolicy change.

However, I am not getting expected behavior which I was expecting for global deny all, hence confused as to why this might be happening. Even without any CNP to allow traffic, I still see traffic reaching app and confused as to why global CCNP is not dropping it with kube-proxy disabled.

### Cilium Version

Client: 1.9.6 38dd27a 2021-04-20T14:57:11-07:00 go version go1.15.11 linux/amd64
Daemon: 1.9.6 38dd27a 2021-04-20T14:57:11-07:00 go version go1.15.11 linux/amd64

### Kernel Version

5.4.158-5.4.7-amd64

### Kubernetes Version

Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.12+rke2r1", GitCommit:"4bf2e32bb2b9fdeea19ff7cdc1fb51fb295ec407", GitTreeState:"clean", BuildDate:"2021-10-28T16:49:46Z", GoVersion:"go1.15.15b5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.12+rke2r1", GitCommit:"4bf2e32bb2b9fdeea19ff7cdc1fb51fb295ec407", GitTreeState:"clean", BuildDate:"2021-10-28T16:49:46Z", GoVersion:"go1.15.15b5", Compiler:"gc", Platform:"linux/amd64"}

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by brb**
See commit msgs.

Fix #19010 

**For v1.10 backport**: feel free to drop tests, as many merge conflicts are expected and we will have a good coverage in the v1.11/main.

### CODE REVIEW

**Review by borkmann — APPROVED** [review]

**Comment by borkmann** [comment]
Short q: If you use kube-proxy, does the remote backend node see WORLD_ID as well for this traffic (or HOST_ID)?

**Review by tklauser — APPROVED** [review]

**Comment by brb** [comment]
> Short q: If you use kube-proxy, does the remote backend node see WORLD_ID as well for this traffic (or HOST_ID)?

I validated it manually - when running with kube-proxy w/o kube-proxy-replacement, the bpf_host @ on cilium_host sets the correct sec ID (i.e., WORLD_ID).

**Comment by brb** [comment]
/test-1.23-net-next

File: test/k8s/services.go [hunk 1]
```diff
@@ -574,6 +574,57 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`,
 			testCurlFromOutsideWithLocalPort(kubectl, ni, svc2URL, 1, false, 64002)
 		})
 
+		It("Tests security id propagation in N/S LB requests fwd-ed over tunnel", func() {
+			// This test case checks whether the "wold" identity is passed in
```
Comment (christarazi) [inline test/k8s/services.go:578]: typo: `wold`

**Review by christarazi — APPROVED** [review]

**Review by joestringer — COMMENTED** [review]
I think that this makes the behaviour more consistent with tunneling mode with kube-proxy, so I like this proposal for that reason. At the same time, I believe that this makes an explicit divide between how tunnel mode handles this case vs. how direct-routing mode handles the case, where direct-routing mode cannot know/transfer this information and hence will always consider such nodeport traffic from other nodes to be 'remote-node', whereas in tunneling environments Cilium would know better.

Any time we change a security identity for something, we need to be a bit careful about how we handle the communication / upgrade story. Even if users today are not consciously allowing traffic from outside the cluster using a `fromEntities: - cluster` / `remote-node` policy, it's possible that they are relying on that policy behaviour. Changing this may force them to re-evaluate which policies they have installed. This could cause production impact upon upgrade.

For this reason, I'm a bit nervous - even wondering whether it's something we need a (hidden) flag for, to mitigate upgrade issues. 

One concrete step we should at least do is to release-note this clearly, something like `Do not allow nodeport traffic via "cluster" fromEntities policy in tunneling mode with KPR`. We probably want to expand on this a bit more to help to highlight that users will need to be paying close attention to their policies for nodeports when they upgrade. What kinds of policies would be affected? What is the recommended way to solve policy drops if you upgrade & find that nodeports don't work any more? etc.

I even wonder if this should _not_ be backported so far in order to reduce the impact for users on earlier Cilium versions, in case they pick up the new micro/patch releases and do not realise it has this impact.

**Comment by brb** [comment]
> At the same time, I believe that this makes an explicit divide between how tunnel mode handles this case vs. how direct-routing mode handles the case, where direct-routing mode cannot know/transfer this information and hence will always consider such nodeport traffic from other nodes to be 'remote-node', whereas in tunneling environments Cilium would know better.

@joestringer Good points. My next step is to better inform users in the documentation (in the KPR and the netpol pages) about the diff, and also to present how the identity can be propagated in the direct-routing mode (i.e., DSR and externalTrafficPolicy=Local).

> Any time we change a security identity for something, we need to be a bit careful about how we handle the communication / upgrade story. Even if users today are not consciously allowing traffic from outside the cluster using a fromEntities: - cluster / remote-node policy, it's possible that they are relying on that policy behaviour. Changing this may force them to re-evaluate which policies they have installed. This could cause production impact upon upgrade.

Maybe we should indeed drop the needs backport label, make the change only in v1.12, and then clearly communicate in the upgrade guide about the change?

> For this reason, I'm a bit nervous - even wondering whether it's something we need a (hidden) flag for, to mitigate upgrade issues.

I'd consider adding the new flag as the last resort :-) 

> One concrete step we should at least do is to release-note this clearly, something like Do not allow nodeport traffic via "cluster" fromEntities policy in tunneling mode with KPR. We probably want to expand on this a bit more to help to highlight that users will need to be paying close attention to their policies for nodeports when they upgrade. What kinds of policies would be affected? What is the recommended way to solve policy drops if you upgrade & find that nodeports don't work any more? etc.

Yeah, ACK. The documentation is planned as the follow-up. I just wanted to get it merged quicker, so that users could start testing. My gut feeling is that the doc changes might require a few reiterations.

**Comment by brb** [comment]
/test-1.21-5.4

**Review by pchaigno — APPROVED** [review]

**Pull request merged**